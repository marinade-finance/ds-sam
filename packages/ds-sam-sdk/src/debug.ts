export class Debug {
  private infos: [string, string][] = []
  private events: string[] = []

  constructor (private readonly voteAccounts: Set<string>) { }

  getVoteAccounts () {
    return this.voteAccounts
  }

  pushInfo (context: string, info: string) {
    this.infos.push([context, info])
  }

  pushValidatorInfo (voteAccount: string, context: string, info: string) {
    if (this.voteAccounts.has(voteAccount)) {
      this.infos.push([context, `${voteAccount} ${info}`])
    }
  }

  pushEvent (event: string) {
    this.events.push(event)
  }

  pushValidatorEvent (voteAccount: string, event: string) {
    if (this.voteAccounts.has(voteAccount)) {
      this.events.push(`${voteAccount} ${event}`)
    }
  }

  pushValidatorSetEvent (voteAccounts: Set<string>, event: string) {
    this.voteAccounts.forEach(debugVoteAccount => {
      if (voteAccounts.has(debugVoteAccount)) {
        this.pushValidatorEvent(debugVoteAccount, event)
      }
    })
  }

  formatInfo (): string {
    return this.infos.map(([context, info]) => `DEBUG INFO - ${context}: ${JSON.stringify(info)}`).join('\n')
  }

  formatEvents (): string {
    return this.events.map(event => `DEBUG EVENT - ${event}`).join('\n')
  }
}
