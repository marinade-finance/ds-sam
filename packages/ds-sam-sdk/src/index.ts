export * from './config'
export * from './sdk'
export * from './types'
export * from './data-provider/data-provider.dto'
export * from './auction'
export * from './debug'
export * from './constraints'
