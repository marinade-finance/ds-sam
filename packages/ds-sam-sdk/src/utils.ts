import { RevShare, Rewards, AggregatedValidator, AuctionValidator, AuctionConstraint, AuctionConstraintType } from './types'

export const MNDE_VOTE_DELEGATION_STRATEGY = 'MarinadeA1gorithmicDe1egationStrategy111111'

export const calcValidatorRevShare = (validator: AggregatedValidator, rewards: Rewards): RevShare => {
  const inflationPmpe = Math.max(0, rewards.inflationPmpe * (1 - validator.inflationCommissionDec))
  const mevPmpe = Math.max(0, rewards.mevPmpe * (1 - (validator.mevCommissionDec ?? 1)))
  const bidPmpe = Math.max(0, validator.bidCpmpe ?? 0)

  return { totalPmpe: inflationPmpe + mevPmpe + bidPmpe, inflationPmpe, mevPmpe, bidPmpe, auctionEffectiveBidPmpe: NaN }
}

export const ineligibleValidatorAggDefaults = () => ({ samEligible: false, mndeEligible: false, ...validatorAggDefaults() })

export const validatorAggDefaults = () => ({ lastCapConstraint: null, stakePriority: NaN, unstakePriority: NaN })

export const validatorTotalAuctionStakeSol = (validator: AuctionValidator): number =>
  validator.auctionStake.externalActivatedSol + validator.auctionStake.marinadeMndeTargetSol + validator.auctionStake.marinadeSamTargetSol

export const zeroStakeConcentration = (type: AuctionConstraintType, name: string, caps: { totalSol: number, marinadeSol: number }): AuctionConstraint => ({
  constraintType: type,
  constraintName: name,
  totalStakeSol: 0,
  totalLeftToCapSol: caps.totalSol,
  marinadeStakeSol: 0,
  marinadeLeftToCapSol: caps.marinadeSol,
  validators: [],
})

export const minCapFromConstraint = (constraint: AuctionConstraint, voteAccounts: Set<string>): { cap: number, affectedValidators: number } => {
  const affectedValidators = constraint.validators.reduce((sum, { voteAccount }) => voteAccounts.has(voteAccount) ? sum + 1 : sum, 0)
  return {
    affectedValidators,
    cap: Math.max(0, Math.min(constraint.totalLeftToCapSol, constraint.marinadeLeftToCapSol)) / affectedValidators
  }
}

export const formatLastCapConstraint = (constraint: AuctionConstraint | null) => constraint ? `${constraint.constraintType} (${constraint.constraintName})` : 'NULL'
