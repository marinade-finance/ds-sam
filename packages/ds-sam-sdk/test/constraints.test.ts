import { DsSamSDK } from '../src'
import { defaultStaticDataProviderBuilder } from './helpers/static-data-provider-builder'
import { ValidatorMockBuilder, generateIdentities, generateVoteAccounts } from './helpers/validator-mock-builder'
import { findValidatorInResult, prettyPrintAuctionResult } from './helpers/utils'

describe('constraints', () => {
  it('applies bond constraints to SAM stake', async () => {
    const voteAccounts = generateVoteAccounts()
    const identities = generateIdentities()

    const validators = [
      new ValidatorMockBuilder(voteAccounts.next().value, identities.next().value)
        .withEligibleDefaults()
        .withMndeVotes(0)
        .withBond({ stakeWanted: 15_000, cpmpe: 0, balance: 1_000 }),
      new ValidatorMockBuilder(voteAccounts.next().value, identities.next().value)
        .withEligibleDefaults()
        .withMndeVotes(0)
        .withBond({ stakeWanted: 20_000, cpmpe: 0, balance: 10 }),
      ...Array.from({ length: 18 }, () => new ValidatorMockBuilder(voteAccounts.next().value, identities.next().value)
        .withEligibleDefaults()
        .withMndeVotes(0)
        .withBond({ stakeWanted: 10_000_000, cpmpe: 0, balance: 1_000 })),
    ]
    const dsSam = new DsSamSDK({}, defaultStaticDataProviderBuilder(validators))
    const result = await dsSam.run()

    expect(prettyPrintAuctionResult(result)).toMatchSnapshot()
  })

  it('applies bond constraints and ignores max stake wanted for MNDE stake', async () => {
    const voteAccounts = generateVoteAccounts()
    const identities = generateIdentities()

    const validators = [
      new ValidatorMockBuilder(voteAccounts.next().value, identities.next().value)
        .withEligibleDefaults()
        .withMndeVotes(2000)
        .withBond({ stakeWanted: 11_000, cpmpe: 0, balance: 20 }),
      new ValidatorMockBuilder(voteAccounts.next().value, identities.next().value)
        .withEligibleDefaults()
        .withMndeVotes(3000)
        .withBond({ stakeWanted: 11_000, cpmpe: 0, balance: 0.1 }),
      ...Array.from({ length: 18 }, () => new ValidatorMockBuilder(voteAccounts.next().value, identities.next().value)
        .withEligibleDefaults()
        .withMndeVotes(1000)
        .withBond({ stakeWanted: 11_000, cpmpe: 0, balance: 1_000 })),
    ]
    const dsSam = new DsSamSDK({}, defaultStaticDataProviderBuilder(validators))
    const result = await dsSam.run()

    expect(prettyPrintAuctionResult(result)).toMatchSnapshot()
  })

  it('applies Marinade stake concentration constraints', async () => {
    const voteAccounts = generateVoteAccounts()
    const identities = generateIdentities()
    const country = 'dummy-country'
    const aso = 'dummy-aso'

    const validators = [
      ...Array.from({ length: 10 }, (_, idx) => new ValidatorMockBuilder(`country-validator-${idx}`, identities.next().value)
        .withEligibleDefaults()
        .withCountry(country)
        .withExternalStake(1_000_000)
        .withBond({ stakeWanted: 1_000_000, cpmpe: 1, balance: 10_000 })),
      ...Array.from({ length: 10 }, (_, idx) => new ValidatorMockBuilder(`country-aso-validator-${idx}`, identities.next().value)
        .withEligibleDefaults()
        .withCountry(country)
        .withAso(aso)
        .withExternalStake(1_000_000)
        .withBond({ stakeWanted: 1_000_000, cpmpe: 1, balance: 10_000 })),
      ...Array.from({ length: 10 }, (_, idx) => new ValidatorMockBuilder(`aso-validator-${idx}`, identities.next().value)
        .withEligibleDefaults()
        .withAso(aso)
        .withExternalStake(1_000_000)
        .withBond({ stakeWanted: 1_000_000, cpmpe: 1, balance: 10_000 })),
      ...Array.from({ length: 20 }, () => new ValidatorMockBuilder(voteAccounts.next().value, identities.next().value)
        .withEligibleDefaults()
        .withMndeVotes(0)
        .withBond({ stakeWanted: 1_000_000, cpmpe: 1, balance: 10_000 })
        .withExternalStake(10_000_000))
    ]
    const dsSam = new DsSamSDK({}, defaultStaticDataProviderBuilder(validators))
    const result = await dsSam.run()

    const countryExternalStake = result.auctionData.validators.reduce((sum, validator) => validator.country === country ? sum + validator.totalActivatedStakeSol : sum, 0)
    const asoExternalStake = result.auctionData.validators.reduce((sum, validator) => validator.aso === aso ? sum + validator.totalActivatedStakeSol : sum, 0)
    const countryMarinadeStake = result.auctionData.validators.reduce((sum, validator) => validator.country === country ? sum + validator.auctionStake.marinadeSamTargetSol + validator.auctionStake.marinadeMndeTargetSol : sum, 0)
    const asoMarinadeStake = result.auctionData.validators.reduce((sum, validator) => validator.aso === aso ? sum + validator.auctionStake.marinadeSamTargetSol + validator.auctionStake.marinadeMndeTargetSol : sum, 0)

    expect(countryMarinadeStake + countryExternalStake).toBeLessThan(result.auctionData.stakeAmounts.networkTotalSol * 0.3)
    expect(asoMarinadeStake + asoExternalStake).toBeLessThan(result.auctionData.stakeAmounts.networkTotalSol * 0.3)
    // Currently Mariande stake country and aso constraints are disabled
    expect(countryMarinadeStake).toBeLessThanOrEqual((result.auctionData.stakeAmounts.marinadeMndeTvlSol + result.auctionData.stakeAmounts.marinadeSamTvlSol) * 1)
    expect(asoMarinadeStake).toBeLessThanOrEqual((result.auctionData.stakeAmounts.marinadeMndeTvlSol + result.auctionData.stakeAmounts.marinadeSamTvlSol) * 1)
    expect(prettyPrintAuctionResult(result)).toMatchSnapshot()
  })

  it('applies global stake concentration constraints', async () => {
    const voteAccounts = generateVoteAccounts()
    const identities = generateIdentities()
    const country = 'dummy-country'
    const aso = 'dummy-aso'

    const validators = [
      ...Array.from({ length: 20 }, (_, idx) => new ValidatorMockBuilder(`country-validator-${idx}`, identities.next().value)
        .withEligibleDefaults()
        .withMndeVotes(0)
        .withCountry(country)
        .withNativeStake(0)
        .withLiquidStake(0)
        .withExternalStake(1_490_000)
        .withBond({ stakeWanted: 1_000_000, cpmpe: 1, balance: 10_000 })),
      ...Array.from({ length: 20 }, (_, idx) => new ValidatorMockBuilder(`country-aso-validator-${idx}`, identities.next().value)
        .withEligibleDefaults()
        .withMndeVotes(0)
        .withCountry(country)
        .withAso(aso)
        .withNativeStake(0)
        .withLiquidStake(0)
        .withExternalStake(1_490_000)
        .withBond({ stakeWanted: 1_000_000, cpmpe: 1, balance: 10_000 })),
      ...Array.from({ length: 20 }, (_, idx) => new ValidatorMockBuilder(`aso-validator-${idx}`, identities.next().value)
        .withEligibleDefaults()
        .withMndeVotes(0)
        .withAso(aso)
        .withNativeStake(0)
        .withLiquidStake(0)
        .withExternalStake(2_000_000)
        .withBond({ stakeWanted: 1_000_000, cpmpe: 1, balance: 1_000_000 })),
      ...Array.from({ length: 20 }, () => new ValidatorMockBuilder(voteAccounts.next().value, identities.next().value)
        .withEligibleDefaults()
        .withMndeVotes(0)
        .withNativeStake(0)
        .withLiquidStake(3_000_000)
        .withExternalStake(5_520_000)),
    ]
    const dsSam = new DsSamSDK({}, defaultStaticDataProviderBuilder(validators))
    const result = await dsSam.run()

    const countryExternalStake = result.auctionData.validators.reduce((sum, validator) => validator.country === country ? sum + validator.totalActivatedStakeSol : sum, 0)
    const asoExternalStake = result.auctionData.validators.reduce((sum, validator) => validator.aso === aso ? sum + validator.totalActivatedStakeSol : sum, 0)
    const countryMarinadeStake = result.auctionData.validators.reduce((sum, validator) => validator.country === country ? sum + validator.auctionStake.marinadeSamTargetSol : sum, 0)
    const asoMarinadeStake = result.auctionData.validators.reduce((sum, validator) => validator.aso === aso ? sum + validator.auctionStake.marinadeSamTargetSol : sum, 0)

    expect(countryMarinadeStake + countryExternalStake).toBeLessThanOrEqual(result.auctionData.stakeAmounts.networkTotalSol * 0.3)
    expect(asoMarinadeStake + asoExternalStake).toBeLessThanOrEqual(result.auctionData.stakeAmounts.networkTotalSol * 0.3)
    // Currently Marinade country and aso constraints are disabled
    expect(countryMarinadeStake).toBeLessThan(result.auctionData.stakeAmounts.marinadeSamTvlSol * 1)
    expect(asoMarinadeStake).toBeLessThan(result.auctionData.stakeAmounts.marinadeSamTvlSol * 1)
    expect(prettyPrintAuctionResult(result)).toMatchSnapshot()
  })

  it('applies validator stake concentration constraints', async () => {
    const voteAccounts = generateVoteAccounts()
    const identities = generateIdentities()

    const validators = [
      new ValidatorMockBuilder('dummy-validator', identities.next().value)
        .withEligibleDefaults()
        .withNativeStake(0)
        .withLiquidStake(0)
        .withExternalStake(100_000)
        .withBond({ stakeWanted: 1_000_000, cpmpe: 1, balance: 10_000 }),
      ...Array.from({ length: 9 }, () => new ValidatorMockBuilder(voteAccounts.next().value, identities.next().value)
        .withEligibleDefaults()
        .withNativeStake(0)
        .withLiquidStake(0)
        .withExternalStake(100_000)
        .withBond({ stakeWanted: 1_000_000, cpmpe: 1, balance: 1 })),
      ...Array.from({ length: 10 }, () => new ValidatorMockBuilder(voteAccounts.next().value, identities.next().value)
        .withEligibleDefaults()
        .withNativeStake(0)
        .withLiquidStake(10_000)
        .withExternalStake(10_000_000)
        .withBond({ stakeWanted: 1_000_000, cpmpe: 0, balance: 1 })),
    ]
    const dsSam = new DsSamSDK({}, defaultStaticDataProviderBuilder(validators))
    const result = await dsSam.run()

    const { auctionStake } = findValidatorInResult('dummy-validator', result)!

    // 100_000 * 0.04 -> TVL * Default cap per validator
    // (0.1 * 100 / 2000) * 100_000 -> (mndeStakeCapMultiplier * validator MNDE votes / total MNDE votes) * TVL
    expect(auctionStake.marinadeSamTargetSol).toStrictEqual(100_000 * 0.04 + (0.1 * 100 / 2000) * 100_000)
    expect(prettyPrintAuctionResult(result)).toMatchSnapshot()
  })
})
