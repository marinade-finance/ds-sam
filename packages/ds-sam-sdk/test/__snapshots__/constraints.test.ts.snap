// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constraints applies validator stake concentration constraints 1`] = `
"dummy-validator, inflation pmpe: 0.4698318496538081, mev pmpe: 0.3956478733926804, bid pmpe: 1, total pmpe: 1.8654797230464886, mnde_target_stake: 0, sam_target_stake: 4500, last constraint: VALIDATOR (dummy-validator)
vote-acc--0000000000, inflation pmpe: 0.4698318496538081, mev pmpe: 0.3956478733926804, bid pmpe: 1, total pmpe: 1.8654797230464886, mnde_target_stake: 0, sam_target_stake: 348.98170521228855, last constraint: BOND (vote-acc--0000000000)
vote-acc--0000000001, inflation pmpe: 0.4698318496538081, mev pmpe: 0.3956478733926804, bid pmpe: 1, total pmpe: 1.8654797230464886, mnde_target_stake: 0, sam_target_stake: 348.98170521228855, last constraint: BOND (vote-acc--0000000001)
vote-acc--0000000002, inflation pmpe: 0.4698318496538081, mev pmpe: 0.3956478733926804, bid pmpe: 1, total pmpe: 1.8654797230464886, mnde_target_stake: 0, sam_target_stake: 348.98170521228855, last constraint: BOND (vote-acc--0000000002)
vote-acc--0000000003, inflation pmpe: 0.4698318496538081, mev pmpe: 0.3956478733926804, bid pmpe: 1, total pmpe: 1.8654797230464886, mnde_target_stake: 0, sam_target_stake: 348.98170521228855, last constraint: BOND (vote-acc--0000000003)
vote-acc--0000000004, inflation pmpe: 0.4698318496538081, mev pmpe: 0.3956478733926804, bid pmpe: 1, total pmpe: 1.8654797230464886, mnde_target_stake: 0, sam_target_stake: 348.98170521228855, last constraint: BOND (vote-acc--0000000004)
vote-acc--0000000005, inflation pmpe: 0.4698318496538081, mev pmpe: 0.3956478733926804, bid pmpe: 1, total pmpe: 1.8654797230464886, mnde_target_stake: 0, sam_target_stake: 348.98170521228855, last constraint: BOND (vote-acc--0000000005)
vote-acc--0000000006, inflation pmpe: 0.4698318496538081, mev pmpe: 0.3956478733926804, bid pmpe: 1, total pmpe: 1.8654797230464886, mnde_target_stake: 0, sam_target_stake: 348.98170521228855, last constraint: BOND (vote-acc--0000000006)
vote-acc--0000000007, inflation pmpe: 0.4698318496538081, mev pmpe: 0.3956478733926804, bid pmpe: 1, total pmpe: 1.8654797230464886, mnde_target_stake: 0, sam_target_stake: 348.98170521228855, last constraint: BOND (vote-acc--0000000007)
vote-acc--0000000008, inflation pmpe: 0.4698318496538081, mev pmpe: 0.3956478733926804, bid pmpe: 1, total pmpe: 1.8654797230464886, mnde_target_stake: 0, sam_target_stake: 348.98170521228855, last constraint: BOND (vote-acc--0000000008)
vote-acc--0000000009, inflation pmpe: 0.4698318496538081, mev pmpe: 0.3956478733926804, bid pmpe: 0, total pmpe: 0.8654797230464886, mnde_target_stake: 0, sam_target_stake: 1155.4285714285713, last constraint: BOND (vote-acc--0000000009)
vote-acc--0000000010, inflation pmpe: 0.4698318496538081, mev pmpe: 0.3956478733926804, bid pmpe: 0, total pmpe: 0.8654797230464886, mnde_target_stake: 0, sam_target_stake: 1155.4285714285713, last constraint: BOND (vote-acc--0000000010)
vote-acc--0000000011, inflation pmpe: 0.4698318496538081, mev pmpe: 0.3956478733926804, bid pmpe: 0, total pmpe: 0.8654797230464886, mnde_target_stake: 0, sam_target_stake: 1155.4285714285713, last constraint: BOND (vote-acc--0000000011)
vote-acc--0000000012, inflation pmpe: 0.4698318496538081, mev pmpe: 0.3956478733926804, bid pmpe: 0, total pmpe: 0.8654797230464886, mnde_target_stake: 0, sam_target_stake: 1155.4285714285713, last constraint: BOND (vote-acc--0000000012)
vote-acc--0000000013, inflation pmpe: 0.4698318496538081, mev pmpe: 0.3956478733926804, bid pmpe: 0, total pmpe: 0.8654797230464886, mnde_target_stake: 0, sam_target_stake: 1155.4285714285713, last constraint: BOND (vote-acc--0000000013)
vote-acc--0000000014, inflation pmpe: 0.4698318496538081, mev pmpe: 0.3956478733926804, bid pmpe: 0, total pmpe: 0.8654797230464886, mnde_target_stake: 0, sam_target_stake: 1155.4285714285713, last constraint: BOND (vote-acc--0000000014)
vote-acc--0000000015, inflation pmpe: 0.4698318496538081, mev pmpe: 0.3956478733926804, bid pmpe: 0, total pmpe: 0.8654797230464886, mnde_target_stake: 0, sam_target_stake: 1155.4285714285713, last constraint: BOND (vote-acc--0000000015)
vote-acc--0000000016, inflation pmpe: 0.4698318496538081, mev pmpe: 0.3956478733926804, bid pmpe: 0, total pmpe: 0.8654797230464886, mnde_target_stake: 0, sam_target_stake: 1155.4285714285713, last constraint: BOND (vote-acc--0000000016)
vote-acc--0000000017, inflation pmpe: 0.4698318496538081, mev pmpe: 0.3956478733926804, bid pmpe: 0, total pmpe: 0.8654797230464886, mnde_target_stake: 0, sam_target_stake: 1155.4285714285713, last constraint: BOND (vote-acc--0000000017)
vote-acc--0000000018, inflation pmpe: 0.4698318496538081, mev pmpe: 0.3956478733926804, bid pmpe: 0, total pmpe: 0.8654797230464886, mnde_target_stake: 0, sam_target_stake: 1155.4285714285713, last constraint: BOND (vote-acc--0000000018)"
`;

exports[`constraints respects bond constraint relaxation I 1`] = `
"vote-acc--0000000000, inflation pmpe: 0.1304586652018676, mev pmpe: 0.10985992859104639, bid pmpe: 100, total pmpe: 100.24031859379292, mnde_target_stake: 0, sam_target_stake: 16763.351749539604, last constraint: BOND (vote-acc--0000000000)
vote-acc--0000000001, inflation pmpe: 0.1304586652018676, mev pmpe: 0.10985992859104639, bid pmpe: 3, total pmpe: 3.2403185937929138, mnde_target_stake: 0, sam_target_stake: 16763.351749539604, last constraint: BOND (vote-acc--0000000001)
vote-acc--0000000002, inflation pmpe: 0.1304586652018676, mev pmpe: 0.10985992859104639, bid pmpe: 2, total pmpe: 2.2403185937929138, mnde_target_stake: 0, sam_target_stake: 23583.133622643953, last constraint: BOND (vote-acc--0000000002)
vote-acc--0000000003, inflation pmpe: 0.1304586652018676, mev pmpe: 0.10985992859104639, bid pmpe: 1, total pmpe: 1.240318593792914, mnde_target_stake: 0, sam_target_stake: 44636.508520289324, last constraint: BOND (vote-acc--0000000003)
vote-acc--0000000004, inflation pmpe: 0.1304586652018676, mev pmpe: 0.10985992859104639, bid pmpe: 0.5, total pmpe: 0.740318593792914, mnde_target_stake: 0, sam_target_stake: 8062.444641275466, last constraint: BOND (vote-acc--0000000004)
vote-acc--0000000005, inflation pmpe: 0.1304586652018676, mev pmpe: 0.10985992859104639, bid pmpe: 0.5, total pmpe: 0.740318593792914, mnde_target_stake: 0, sam_target_stake: 8062.444641275466, last constraint: BOND (vote-acc--0000000005)
vote-acc--0000000006, inflation pmpe: 0.1304586652018676, mev pmpe: 0.10985992859104639, bid pmpe: 0.5, total pmpe: 0.740318593792914, mnde_target_stake: 0, sam_target_stake: 8062.444641275466, last constraint: BOND (vote-acc--0000000006)
vote-acc--0000000007, inflation pmpe: 0.1304586652018676, mev pmpe: 0.10985992859104639, bid pmpe: 0.5, total pmpe: 0.740318593792914, mnde_target_stake: 0, sam_target_stake: 8062.444641275466, last constraint: BOND (vote-acc--0000000007)
vote-acc--0000000008, inflation pmpe: 0.1304586652018676, mev pmpe: 0.10985992859104639, bid pmpe: 0.5, total pmpe: 0.740318593792914, mnde_target_stake: 0, sam_target_stake: 8062.444641275466, last constraint: BOND (vote-acc--0000000008)
vote-acc--0000000009, inflation pmpe: 0.1304586652018676, mev pmpe: 0.10985992859104639, bid pmpe: 0.5, total pmpe: 0.740318593792914, mnde_target_stake: 0, sam_target_stake: 8062.444641275466, last constraint: BOND (vote-acc--0000000009)
vote-acc--0000000010, inflation pmpe: 0.1304586652018676, mev pmpe: 0.10985992859104639, bid pmpe: 0.5, total pmpe: 0.740318593792914, mnde_target_stake: 0, sam_target_stake: 8062.444641275466, last constraint: BOND (vote-acc--0000000010)
vote-acc--0000000011, inflation pmpe: 0.1304586652018676, mev pmpe: 0.10985992859104639, bid pmpe: 0.5, total pmpe: 0.740318593792914, mnde_target_stake: 0, sam_target_stake: 8062.444641275466, last constraint: BOND (vote-acc--0000000011)
vote-acc--0000000012, inflation pmpe: 0.1304586652018676, mev pmpe: 0.10985992859104639, bid pmpe: 0.5, total pmpe: 0.740318593792914, mnde_target_stake: 0, sam_target_stake: 8062.444641275466, last constraint: BOND (vote-acc--0000000012)
vote-acc--0000000013, inflation pmpe: 0.1304586652018676, mev pmpe: 0.10985992859104639, bid pmpe: 0.5, total pmpe: 0.740318593792914, mnde_target_stake: 0, sam_target_stake: 8062.444641275466, last constraint: BOND (vote-acc--0000000013)
vote-acc--0000000014, inflation pmpe: 0.1304586652018676, mev pmpe: 0.10985992859104639, bid pmpe: 0.5, total pmpe: 0.740318593792914, mnde_target_stake: 0, sam_target_stake: 8062.444641275466, last constraint: BOND (vote-acc--0000000014)
vote-acc--0000000015, inflation pmpe: 0.1304586652018676, mev pmpe: 0.10985992859104639, bid pmpe: 0.5, total pmpe: 0.740318593792914, mnde_target_stake: 0, sam_target_stake: 8062.444641275466, last constraint: BOND (vote-acc--0000000015)
vote-acc--0000000016, inflation pmpe: 0.1304586652018676, mev pmpe: 0.10985992859104639, bid pmpe: 0.5, total pmpe: 0.740318593792914, mnde_target_stake: 0, sam_target_stake: 8062.444641275466, last constraint: BOND (vote-acc--0000000016)
vote-acc--0000000017, inflation pmpe: 0.1304586652018676, mev pmpe: 0.10985992859104639, bid pmpe: 0.5, total pmpe: 0.740318593792914, mnde_target_stake: 0, sam_target_stake: 8062.444641275466, last constraint: BOND (vote-acc--0000000017)
vote-acc--0000000018, inflation pmpe: 0.1304586652018676, mev pmpe: 0.10985992859104639, bid pmpe: 0.5, total pmpe: 0.740318593792914, mnde_target_stake: 0, sam_target_stake: 8062.444641275466, last constraint: BOND (vote-acc--0000000018)
vote-acc--0000000019, inflation pmpe: 0.1304586652018676, mev pmpe: 0.10985992859104639, bid pmpe: 0.5, total pmpe: 0.740318593792914, mnde_target_stake: 0, sam_target_stake: 8062.444641275466, last constraint: BOND (vote-acc--0000000019)
vote-acc--0000000020, inflation pmpe: 0.1304586652018676, mev pmpe: 0.10985992859104639, bid pmpe: 0.5, total pmpe: 0.740318593792914, mnde_target_stake: 0, sam_target_stake: 8062.444641275466, last constraint: BOND (vote-acc--0000000020)
vote-acc--0000000021, inflation pmpe: 0.1304586652018676, mev pmpe: 0.10985992859104639, bid pmpe: 0.5, total pmpe: 0.740318593792914, mnde_target_stake: 0, sam_target_stake: 8062.444641275466, last constraint: BOND (vote-acc--0000000021)"
`;

exports[`constraints respects bond constraint relaxation II 1`] = `
"vote-acc--0000000000, inflation pmpe: 0.13033337906434353, mev pmpe: 0.10975442447523664, bid pmpe: 100, total pmpe: 100.24008780353958, mnde_target_stake: 0, sam_target_stake: 44175.75757575757, last constraint: BOND (vote-acc--0000000000)
vote-acc--0000000001, inflation pmpe: 0.13033337906434353, mev pmpe: 0.10975442447523664, bid pmpe: 1, total pmpe: 1.2400878035395801, mnde_target_stake: 0, sam_target_stake: 44175.75757575757, last constraint: BOND (vote-acc--0000000001)
vote-acc--0000000002, inflation pmpe: 0.13033337906434353, mev pmpe: 0.10975442447523664, bid pmpe: 0.5, total pmpe: 0.7400878035395801, mnde_target_stake: 0, sam_target_stake: 44175.75757575757, last constraint: BOND (vote-acc--0000000002)
vote-acc--0000000005, inflation pmpe: 0.13033337906434353, mev pmpe: 0.10975442447523664, bid pmpe: 0.5, total pmpe: 0.7400878035395801, mnde_target_stake: 0, sam_target_stake: 8835.151515151514, last constraint: BOND (vote-acc--0000000005)
vote-acc--0000000006, inflation pmpe: 0.13033337906434353, mev pmpe: 0.10975442447523664, bid pmpe: 0.5, total pmpe: 0.7400878035395801, mnde_target_stake: 0, sam_target_stake: 8835.151515151514, last constraint: BOND (vote-acc--0000000006)
vote-acc--0000000007, inflation pmpe: 0.13033337906434353, mev pmpe: 0.10975442447523664, bid pmpe: 0.5, total pmpe: 0.7400878035395801, mnde_target_stake: 0, sam_target_stake: 8835.151515151514, last constraint: BOND (vote-acc--0000000007)
vote-acc--0000000008, inflation pmpe: 0.13033337906434353, mev pmpe: 0.10975442447523664, bid pmpe: 0.5, total pmpe: 0.7400878035395801, mnde_target_stake: 0, sam_target_stake: 8835.151515151514, last constraint: BOND (vote-acc--0000000008)
vote-acc--0000000009, inflation pmpe: 0.13033337906434353, mev pmpe: 0.10975442447523664, bid pmpe: 0.5, total pmpe: 0.7400878035395801, mnde_target_stake: 0, sam_target_stake: 8835.151515151514, last constraint: BOND (vote-acc--0000000009)
vote-acc--0000000010, inflation pmpe: 0.13033337906434353, mev pmpe: 0.10975442447523664, bid pmpe: 0.5, total pmpe: 0.7400878035395801, mnde_target_stake: 0, sam_target_stake: 8835.151515151514, last constraint: BOND (vote-acc--0000000010)
vote-acc--0000000011, inflation pmpe: 0.13033337906434353, mev pmpe: 0.10975442447523664, bid pmpe: 0.5, total pmpe: 0.7400878035395801, mnde_target_stake: 0, sam_target_stake: 8835.151515151514, last constraint: BOND (vote-acc--0000000011)
vote-acc--0000000012, inflation pmpe: 0.13033337906434353, mev pmpe: 0.10975442447523664, bid pmpe: 0.5, total pmpe: 0.7400878035395801, mnde_target_stake: 0, sam_target_stake: 8835.151515151514, last constraint: BOND (vote-acc--0000000012)
vote-acc--0000000013, inflation pmpe: 0.13033337906434353, mev pmpe: 0.10975442447523664, bid pmpe: 0.5, total pmpe: 0.7400878035395801, mnde_target_stake: 0, sam_target_stake: 8835.151515151514, last constraint: BOND (vote-acc--0000000013)
vote-acc--0000000014, inflation pmpe: 0.13033337906434353, mev pmpe: 0.10975442447523664, bid pmpe: 0.5, total pmpe: 0.7400878035395801, mnde_target_stake: 0, sam_target_stake: 8835.151515151514, last constraint: BOND (vote-acc--0000000014)
vote-acc--0000000015, inflation pmpe: 0.13033337906434353, mev pmpe: 0.10975442447523664, bid pmpe: 0.5, total pmpe: 0.7400878035395801, mnde_target_stake: 0, sam_target_stake: 8835.151515151514, last constraint: BOND (vote-acc--0000000015)
vote-acc--0000000016, inflation pmpe: 0.13033337906434353, mev pmpe: 0.10975442447523664, bid pmpe: 0.5, total pmpe: 0.7400878035395801, mnde_target_stake: 0, sam_target_stake: 8835.151515151514, last constraint: BOND (vote-acc--0000000016)
vote-acc--0000000017, inflation pmpe: 0.13033337906434353, mev pmpe: 0.10975442447523664, bid pmpe: 0.5, total pmpe: 0.7400878035395801, mnde_target_stake: 0, sam_target_stake: 8835.151515151514, last constraint: BOND (vote-acc--0000000017)
vote-acc--0000000018, inflation pmpe: 0.13033337906434353, mev pmpe: 0.10975442447523664, bid pmpe: 0.5, total pmpe: 0.7400878035395801, mnde_target_stake: 0, sam_target_stake: 8835.151515151514, last constraint: BOND (vote-acc--0000000018)
vote-acc--0000000019, inflation pmpe: 0.13033337906434353, mev pmpe: 0.10975442447523664, bid pmpe: 0.5, total pmpe: 0.7400878035395801, mnde_target_stake: 0, sam_target_stake: 8835.151515151514, last constraint: BOND (vote-acc--0000000019)
vote-acc--0000000020, inflation pmpe: 0.13033337906434353, mev pmpe: 0.10975442447523664, bid pmpe: 0.5, total pmpe: 0.7400878035395801, mnde_target_stake: 0, sam_target_stake: 8835.151515151514, last constraint: BOND (vote-acc--0000000020)
vote-acc--0000000021, inflation pmpe: 0.13033337906434353, mev pmpe: 0.10975442447523664, bid pmpe: 0.5, total pmpe: 0.7400878035395801, mnde_target_stake: 0, sam_target_stake: 8835.151515151514, last constraint: BOND (vote-acc--0000000021)
vote-acc--0000000022, inflation pmpe: 0.13033337906434353, mev pmpe: 0.10975442447523664, bid pmpe: 0.5, total pmpe: 0.7400878035395801, mnde_target_stake: 0, sam_target_stake: 8835.151515151514, last constraint: BOND (vote-acc--0000000022)
vote-acc--0000000003, inflation pmpe: 0.13033337906434353, mev pmpe: 0.10975442447523664, bid pmpe: 0.2, total pmpe: 0.4400878035395802, mnde_target_stake: 0, sam_target_stake: 78114.28326474623, last constraint: BOND (vote-acc--0000000003)
vote-acc--0000000004, inflation pmpe: 0.13033337906434353, mev pmpe: 0.10975442447523664, bid pmpe: 0.1, total pmpe: 0.3400878035395802, mnde_target_stake: 0, sam_target_stake: 113613.69162666, last constraint: BOND (vote-acc--0000000004)"
`;
